lib/logstash/filters/geoip.rb

```diff
@@ -82,139 +86,121 @@ class LogStash::Filters::GeoIP < LogStash::Filters::Base
-    # not set as a global. The geoip module imposes a mutex, so the filter needs
-    # to re-initialize this later in the filter() thread, and save that access
-    # as a thread-local variable.
-    geoip_initialize = ::GeoIP.new(@database)
-
-    @geoip_type = case geoip_initialize.database_type
-    when GeoIP::GEOIP_CITY_EDITION_REV0, GeoIP::GEOIP_CITY_EDITION_REV1
-      :city
-    when GeoIP::GEOIP_COUNTRY_EDITION
-      :country
-    when GeoIP::GEOIP_ASNUM_EDITION
-      :asn
-    when GeoIP::GEOIP_ISP_EDITION, GeoIP::GEOIP_ORG_EDITION
-      :isp
-    else
-      raise RuntimeException.new "This GeoIP database is not currently supported"
-    end
 
-    @threadkey = "geoip-#{self.object_id}"
+    @logger.info("Using geoip database", :path => @database)
 
-    # This is wrapped in a mutex to make sure the initialization behavior of LOOKUP_CACHES (see def above) doesn't create a dupe
-    LOOKUP_CACHE_INIT_MUTEX.synchronize do
-      self.lookup_cache = LOOKUP_CACHES[@geoip_type] ||= LruRedux::ThreadSafeCache.new(1000)
+    db_file = JavaIO::File.new(@database)
+    begin
+      @parser = DatabaseReader::Builder.new(db_file).withCache(CHMCache.new(@cache_size)).build();
+    rescue Java::ComMaxmindDb::InvalidDatabaseException => e
+      # do nothing
     end
-
-    @no_fields = @fields.nil? || @fields.empty?
   end # def register
 
   public
   def filter(event)
-    geo_data_hash = get_geo_data(event)
-    if apply_geodata(geo_data_hash, event)
-      filter_matched(event)
-    end
-  end # def filter
+    return unless filter?(event)
 
-  def apply_geodata(geo_data_hash, event)
-    # don't do anything more if the lookup result is nil?
-    return false if geo_data_hash.nil?
-    # only set the event[@target] if the lookup result is not nil: BWC
-    event[@target] = {} if event[@target].nil?
-    # don't do anything more if the lookup result is empty?
-    return false if geo_data_hash.empty?
-    geo_data_hash.each do |key, value|
-      if @no_fields || @fields.include?(key)
-        # can't dup numerics
-        event["[#{@target}][#{key}]"] = value.is_a?(Numeric) ? value : value.dup
-      end
-    end # geo_data_hash.each
-    true
-  end
-
-  def get_geo_data(event)
-    # pure function, must control return value
-    result = {}
-    ip = event[@source]
-    ip = ip.first if ip.is_a? Array
-    return nil if ip.nil?
     begin
-      result = get_geo_data_for_ip(ip)
-    rescue SocketError => e
+      ip = event[@source]
+      ip = ip.first if ip.is_a? Array
+      ip_address = InetAddress.getByName(ip)
+      response = @parser.city(ip_address)
+      country = response.getCountry()
+      subdivision = response.getMostSpecificSubdivision()
+      city = response.getCity()
+      postal = response.getPostal()
+      location = response.getLocation()
+
+      geo_data_hash = Hash.new()
+
+      if @fields.nil? || @fields.empty? || @fields.include?("city_name")
```

Comment:
instead of re-doing all these checks, here is a commit showing a potential refactor I think would make things more readable

https://github.com/logstash-plugins/logstash-filter-geoip/commit/55abe223aa3e9fe726527629e177dda10f990abe